### PR TITLE
[MU4] New Score Wizard. Instruments page

### PR DESCRIPTION
**Pull request by Eism**
Partially depends on #6454.

Peek 2020-08-14 14-21

### CODE REVIEW

File: mu4/scenes/instruments/view/instrumentlistmodel.cpp [hunk 1]
```diff
@@ -145,12 +162,178 @@ void InstrumentListModel::selectGroup(const QString& group)
...
+        { "g-", "G" }
+    };
+
+    for (const Transposition& transposition: transpositions) {
+        if (instrumentId.startsWith(transposition.id)) {
+            return transposition;
+        }
+    }
+    return Transposition();
+}
```
Comment (igorkorsukov) [inline mu4/scenes/instruments/view/instrumentlistmodel.cpp:271]: I think better move this method to some service class, not viewmodel

File: mu4/scenes/instruments/view/instrumentlistmodel.h [hunk 2]
```diff
@@ -46,25 +47,61 @@ class InstrumentListModel : public QObject, public async::Asyncable
...
 
 private:
+    struct Transposition {
+        QString id;
+        QString name;
+
+        bool isValid()
+        {
+            return !id.isEmpty();
+        }
```
Comment (igorkorsukov) [inline mu4/scenes/instruments/view/instrumentlistmodel.h:80]: Better to make one line and add `const`

File: mu4/userscores/qml/MuseScore/UserScores/ChooseInstrumentsAndTemplatesPage.qml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        contentHeight: 28
+        spacing: 0
+
+        StyledTabButton {
+            text: qsTrc("userscores", "Choose instruments")
+            sideMargin: 22
+            isCurrent: bar.currentIndex === 0
+        }
+        StyledTabButton {
+            text: qsTrc("appshell", "Choose from template")
```
Comment (igorkorsukov) [inline mu4/userscores/qml/MuseScore/UserScores/ChooseInstrumentsAndTemplatesPage.qml:29]: most likely there should be `userscores`

File: mu4/scenes/instruments/qml/MuseScore/Instruments/SelectedInstrumentsView.qml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import MuseScore.Ui 1.0
+import MuseScore.UiComponents 1.0
+import MuseScore.Instruments 1.0
+
+Item {
+    id: root
+
+    property var instruments: null
+
+    property bool isCanUpInstrument: currentInstrumentIndex > 0
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/qml/MuseScore/Instruments/SelectedInstrumentsView.qml:13]: canLiftUpInstument?

File: mu4/scenes/instruments/view/instrumentlistmodel.cpp [hunk 5]
```diff
@@ -109,23 +99,49 @@ QVariantList InstrumentListModel::groups()
...
+            instrumentId = instrumentId.replace(_transposition.id, "");
+            instrumentName = instrumentName.replace(_transposition.name + " ", "")
+                             .replace(" in " + _transposition.name, "");
+
+            if (availableInstruments.contains(instrumentId)) {
+                instrumentTranspositions = availableInstruments[instrumentId].toMap().value("transpositions").toList();
+            }
+
+            QVariantMap obj;
+            obj["id"] = _transposition.id;
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/view/instrumentlistmodel.cpp:128]: These keys ("id", "name") can be found in several places. Better to add constants: static const QString KEY_ID("id")

File: mu4/scenes/instruments/view/instrumentlistmodel.cpp [hunk 6]
```diff
@@ -145,12 +162,183 @@ void InstrumentListModel::selectGroup(const QString& group)
...
+        }
+    } else {
+        if (m_selectedFamilyId != ALL_INSTRUMENTS_ID) {
+            m_savedSelectedFamilyId = m_selectedFamilyId;
+            selectFamily(ALL_INSTRUMENTS_ID);
+        }
+    }
+}
+
+bool InstrumentListModel::isInstrumentAccepted(const InstrumentTemplate &instrument) const
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/view/instrumentlistmodel.cpp:335]: const InstrumentTemplate& instrument

File: mu4/scenes/instruments/qml/MuseScore/Instruments/FamilyView.qml [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+        delegate: Item {
+            width: parent.width
+            height: 40
+
+            Rectangle {
+                anchors.fill: parent
+
+                color: privateProperties.currentGroupIndex === index ? ui.theme.accentColor : ui.theme.backgroundColor
+                opacity: privateProperties.currentGroupIndex === index ? 0.3 : 1
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/qml/MuseScore/Instruments/FamilyView.qml:97]: Seems that variations of this ListView can be found in 3 different places. Better to create a general component and reuse it with different configuration

File: mu4/scenes/instruments/view/instrumentlistmodel.cpp [hunk 8]
```diff
@@ -145,12 +162,183 @@ void InstrumentListModel::selectGroup(const QString& group)
...
+        obj["name"] = instrumentName;
+        result << obj;
+    }
+
+    return result;
+}
+
+InstrumentListModel::Transposition InstrumentListModel::transposition(const QString& instrumentId) const
+{
+    static Transposition transpositions[] = {
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/view/instrumentlistmodel.cpp:256]: Need to implement translations of these strings

File: mu4/scenes/instruments/qml/MuseScore/Instruments/SelectedInstrumentsView.qml [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        anchors.right: parent.right
+
+        StyledComboBox {
+            Layout.fillWidth: true
+
+            textRoleName: "text"
+            valueRoleName: "value"
+
+            model: {
+                // TODO: not implemented
```
Comment (RomanPudashkin) [inline mu4/scenes/instruments/qml/MuseScore/Instruments/SelectedInstrumentsView.qml:46]: Better transfer that code (with TODO) to a model

**Review by RomanPudashkin — APPROVED** [review]

**Pull request merged**